File: exercises/csharp/fill_the_bucket/mutate/FillBucket.cs
using System;
using System.Collections.Generic;
using System.Linq;

// ============================================================
// Fill the bucket - find weak tests with mutation testing
// Goal: fill the big bucket with minimum fill operations
// a fill operation = using a small bucket of any size
// example: using 7 twice + 2 three times = 5 operations
// @Author: Amir Kirsh
// ============================================================

public static class BucketFiller
{
    private static (bool CanFill, Dictionary<int, int> Plan) CanFillRecursive(
        int bigBucket, List<int> smallBuckets, int index)
    {
        if (bigBucket < smallBuckets[^1])
            return (false, new Dictionary<int, int>());
        var curr = smallBuckets[index];
        if (bigBucket % curr == 0)
            return (true, new Dictionary<int, int> { { curr, bigBucket / curr } });
        if (index < smallBuckets.Count - 1)
        {
            var times = bigBucket / curr + 1;
            while (times > 0)
            {
                times--;
                var rest = bigBucket - times * curr;
                var (canFill, fillPlan) = CanFillRecursive(rest, smallBuckets, index + 1);
                if (canFill)
                {
                    if (times > 0) fillPlan[curr] = times;
                    return (true, fillPlan);
                }
            }
        }
        return (false, new Dictionary<int, int>());
    }

    public static (bool CanFill, Dictionary<int, int> Plan) CanFill(
        int bigBucket, IEnumerable<int> smallBuckets)
    {
        if (bigBucket == 0)
            return (true, new Dictionary<int, int>());
        var bucketsList = smallBuckets.ToList();
        if (!bucketsList.Any())
            return (false, new Dictionary<int, int>());
        var bucketsSorted = bucketsList.Distinct().OrderByDescending(x => x).ToList();
        if (bucketsSorted[^1] == 0) bucketsSorted.RemoveAt(bucketsSorted.Count - 1);
        return CanFillRecursive(bigBucket, bucketsSorted, 0);
    }
}

// ============================================================
// Test cases
// ============================================================
using Xunit;

public class FillBucketTests
{
    private static void AssertFillResult(
        bool expectedCanFill, Dictionary<int, int> expectedPlan,
        bool resultCanFill, Dictionary<int, int> resultPlan)
    {
        Assert.Equal(expectedCanFill, resultCanFill);
        Assert.Equal(expectedPlan.Count, resultPlan.Count);
        foreach (var (key, value) in expectedPlan)
        {
            Assert.True(resultPlan.ContainsKey(key), $"Missing bucket size {key}");
            Assert.Equal(value, resultPlan[key]);
        }
    }

    // Full tests - check both boolean and fill plan
    [Theory]
    [MemberData(nameof(CanFillBasicData))]
    public void TestCanFillBasic(int bigBucket, int[] smallBuckets,
        bool expectedCanFill, Dictionary<int, int> expectedPlan)
    {
        var (canFill, plan) = BucketFiller.CanFill(bigBucket, smallBuckets);
        AssertFillResult(expectedCanFill, expectedPlan, canFill, plan);
    }

    public static IEnumerable<object[]> CanFillBasicData()
    {
        yield return new object[] { 10, new[] { 2 }, true, new Dictionary<int, int> { { 2, 5 } } };
        yield return new object[] { 12, new[] { 3, 4 }, true, new Dictionary<int, int> { { 4, 3 } } };
    }

    [Theory]
    [MemberData(nameof(CannotFillBasicData))]
    public void TestCannotFillBasic(int bigBucket, int[] smallBuckets,
        bool expectedCanFill, Dictionary<int, int> expectedPlan)
    {
        var (canFill, plan) = BucketFiller.CanFill(bigBucket, smallBuckets);
        AssertFillResult(expectedCanFill, expectedPlan, canFill, plan);
    }

    public static IEnumerable<object[]> CannotFillBasicData()
    {
        yield return new object[] { 10, new[] { 3 }, false, new Dictionary<int, int>() };
        yield return new object[] { 10, new[] { 13, 11 }, false, new Dictionary<int, int>() };
    }

    // Weak tests - only check the boolean result, not the plan
    [Theory]
    [MemberData(nameof(CanFillMultiBucketData))]
    public void TestCanFillMultiBucket(int bigBucket, int[] smallBuckets, bool expectedCanFill)
    {
        var (canFill, _) = BucketFiller.CanFill(bigBucket, smallBuckets);
        Assert.Equal(expectedCanFill, canFill);
    }

    public static IEnumerable<object[]> CanFillMultiBucketData()
    {
        yield return new object[] { 8, new[] { 3, 5 }, true };
        yield return new object[] { 11, new[] { 4, 3 }, true };
        yield return new object[] { 11, new[] { 5, 3 }, true };
        yield return new object[] { 111, new[] { 4, 5, 3, 200 }, true };
    }

    [Theory]
    [MemberData(nameof(CannotFillMultiBucketData))]
    public void TestCannotFillMultiBucket(int bigBucket, int[] smallBuckets, bool expectedCanFill)
    {
        var (canFill, _) = BucketFiller.CanFill(bigBucket, smallBuckets);
        Assert.Equal(expectedCanFill, canFill);
    }

    public static IEnumerable<object[]> CannotFillMultiBucketData()
    {
        yield return new object[] { 11, new[] { 6, 3 }, false };
    }
}

File: exercises/python/fill_the_bucket/mutate/fill_bucket.py
"""
Fill the bucket - find weak tests with mutation testing
Goal: fill the big bucket with minimum fill operations
a fill operation = using a small bucket of any size
example: using 7 twice + 2 three times = 5 operations
@Author: Amir Kirsh
"""

from typing import List, Dict, Tuple

FillPlan = Dict[int, int]
FillResult = Tuple[bool, FillPlan]


def _can_fill_recursive(big_bucket: int, small_buckets: List[int], index: int) -> FillResult:
    if big_bucket < small_buckets[-1]:
        return False, {}
    curr = small_buckets[index]
    if big_bucket % curr == 0:
        return True, {curr: big_bucket // curr}
    if index < len(small_buckets) - 1:
        times = big_bucket // curr + 1
        while times > 0:
            times -= 1
            rest = big_bucket - times * curr
            can, fill_plan = _can_fill_recursive(rest, small_buckets, index + 1)
            if can:
                if times > 0:
                    fill_plan[curr] = times
                return True, fill_plan
    return False, {}


def can_fill(big_bucket: int, small_buckets) -> FillResult:
    if big_bucket == 0:
        return True, {}
    buckets_list = list(small_buckets)
    if not buckets_list:
        return False, {}
    buckets_sorted = sorted(set(buckets_list), reverse=True)
    if buckets_sorted[-1] == 0:
        buckets_sorted.pop()
    return _can_fill_recursive(big_bucket, buckets_sorted, 0)


# ============================================================
# Test cases
# ============================================================
import pytest


# Full tests - check both boolean and fill plan
@pytest.mark.parametrize("big_bucket, small_buckets, expected_can_fill, expected_plan", [
    (10, [2], True, {2: 5}),
    (12, [3, 4], True, {4: 3}),
])
def test_can_fill_basic(big_bucket, small_buckets, expected_can_fill, expected_plan):
    result_can_fill, result_plan = can_fill(big_bucket, small_buckets)
    assert result_can_fill == expected_can_fill
    assert result_plan == expected_plan


@pytest.mark.parametrize("big_bucket, small_buckets, expected_can_fill, expected_plan", [
    (10, [3], False, {}),
    (10, [13, 11], False, {}),
])
def test_cannot_fill_basic(big_bucket, small_buckets, expected_can_fill, expected_plan):
    result_can_fill, result_plan = can_fill(big_bucket, small_buckets)
    assert result_can_fill == expected_can_fill
    assert result_plan == expected_plan


# Weak tests - only check the boolean result, not the plan
@pytest.mark.parametrize("big_bucket, small_buckets, expected_can_fill", [
    (8, [3, 5], True),
    (11, [4, 3], True),
    (11, [5, 3], True),
    (111, [4, 5, 3, 200], True),
])
def test_can_fill_multi_bucket(big_bucket, small_buckets, expected_can_fill):
    result_can_fill, _ = can_fill(big_bucket, small_buckets)
    assert result_can_fill == expected_can_fill


@pytest.mark.parametrize("big_bucket, small_buckets, expected_can_fill", [
    (11, [6, 3], False),
])
def test_cannot_fill_multi_bucket(big_bucket, small_buckets, expected_can_fill):
    result_can_fill, _ = can_fill(big_bucket, small_buckets)
    assert result_can_fill == expected_can_fill
